README
NAME
    whouses - impact analysis in a clearmake build environment

MOTTO
    *"You give me a clean CR, I'll give you a clean impact analysis."*

SYNOPSIS
    Run this script with the "-help" option for usage details. Here are some
    additional sample usages with explanations:

      whouses foobar.h

    Shows all DO's that make use of any file matching /foobar.h/.

      whouses -recurse foobar.h

    Same as above but follows the chain of derived files recursively.

      whouses -exact foobar.h

    Shows all DO's that make use of the specified file. The "-exact" flag
    suppresses pattern matching and shows only DO's which reference the
    exact file.

DESCRIPTION
    Whouses provides a limited form of "impact analysis" in a clearmake
    build environment. This is different from traditional impact analysis
    (see TRUE CODE ANALYSIS COMPARED below for details). In particular, it
    operates at the granularity of files rather than language elements.

    Whouses is best described by example. Imagine you have a VOB /vobs_sw in
    which you build the incredibly simple application "foo" from "foo.c".
    You have a Makefile which compiles "foo.c" to "foo.o" and then links it
    to produce "foo". And let's further assume you've just done a build
    using clearmake.

    Thus, "foo" is a DO which has a config record (*CR*) showing how it was
    made. Whouses analyzes that CR and prints the data in easy-to-read
    indented textual format. For instance:

            % whouses -do foo foo.c
            /vobs_sw/src/foo.c  =>
              /vobs_sw/src/foo.o

    The "-do foo" points to the derived object from which to extract and
    analyze the CR; it will be implicit in the remaining examples. The
    output indicates that "foo.o uses foo.c", or in other words that "foo.c"
    is a contributor to "foo.o". If we add the "-recurse" flag:

            % whouses -r foo.c
            /vobs_sw/src/foo.c =>
              /vobs_sw/src/foo.o
                /vobs_sw/src/foo

    We see all files to which "foo.c" contributes, indented according to how
    many generations removed they are. If we now add "-terminals"

            % whouses -r -t foo.c
            /vobs_sw/src/foo.c =>
              /vobs_sw/src/foo

    Intermediate targets such as "foo.o" are suppressed so we see only the
    "final" targets descended from "foo.c".

    We can also go in the other direction using "-backwards":

            % whouses -b -e foo
            /vobs_sw/src/foo <=
              /vobs_sw/src/foo.o

    Which shows "foo.o" as a progenitor of "foo". Note that the arrow (<=)
    is turned around to indicate "-backwards" mode. We also introduced the
    "-exact" flag here. By default arguments to whouses are treated as
    patterns, not file names, and since "foo" has no extension it would have
    matched "foo.c" and "foo.o" as well. Use of "-exact" suppresses pattern
    matching.

    We can combine recursion with backwards mode:

            % whouses -back -exact -recurse foo
            /vobs_sw/src/foo <=
              /vobs_sw/src/foo.o
                  /vobs_sw/src/foo.c
                  /vobs_sw/src/foo.h
                  /vobs_sw/src/bar.h

    And discover that "foo.h" and "bar.h" were also used.

    When used recursively in the forward direction, this script answers the
    question "if I change file X, which derived files will need to be
    rebuilt"? This is the classic use, the one for which it was written.
    When used recursively in the backward direction it can depict the entire
    dependency tree in an easily readable format.

    Because extracting a recursive CR can be quite slow for large build
    systems, whouses provides ways of dumping the CR data to a file
    representation for speed. Use of "-save":

            % whouses -do foo -save ...

    will write out the data to foo.crdb. Subsequently, if foo.crdb exists it
    will be used unless a new the "-do" flag is used. See also the "-db" and
    "-fmt" flags.

    The default save format is that of Data::Dumper. It was chosen because
    it results in a nicely indented, human-readable text format file.
    Provision is made for subclassing to alternate storage formats, and
    subclasses for Storable and Data::Denter are provided as examples.
    Although the Storable format may be somewhat faster than the default,
    the real reason for the override capability is to allow an actual, fast
    database to be used.

SELECTING DERIVED OBJECTS
    If a "-do" flag is given, the CRs are taken from the specified derived
    object(s). Multiple DO's may be specified with multiple "-do" flags or
    as a comma-separated list. Alternatively, if the "CRDB_DO" environment
    variable exists, its value is used as if specified with "-do".

    If no DO's are specified directly, "whouses" will look for stored DO
    data in files specified with "-db" or the "CRDB_DB" EV. The format is
    the same as above.

    Failing that, "whouses" will search for files named "*.crdb" along a
    path specified with "-dir" or "CRDB_PATH", defaulting to the current
    directory.

.AUDIT FILES
    As a special case, derived objects matching the Perl regular expression
    "/\.AUDIT/i" are ignored while traversing the recursive config spec.
    These are presumed to be *meta-DO's* by convention, which aren't part of
    the build per se but rather pseudo-targets whose only purpose is to hold
    CR's which refer back to all real deliverables. Thus if you construct
    your Makefile to create a meta-DO, you might want to name it ".AUDIT" or
    ".prog.AUDIT" or something.

ClearCase::CRDB
    Most of the logic is actually in the "ClearCase::CRDB" module; the
    "whouses" program is just a wrapper that uses the module. It's done this
    way so ClearCase::CRDB can provide an API for other potential tools
    which need to do CR analysis.

TRUE CODE ANALYSIS COMPARED
    Whouses can best be described by comparison with "real" impact analysis
    products. There are a number of such tools on the market, for example
    SNiFF+ from WindRiver. Typically these work by parsing the source code
    into some database representation which they can then analyze. It's a
    powerful technique but entails some tradeoffs:

  MINUSES
    *   A true code analysis tool must have knowledge of each programming
        language in use. I.e. to add support for Java, a Java parser must be
        added.

    *   A corollary of the above is that this tool requires lot of work by
        expert programmers. Thus the tools tend to be large, complex and
        expensive. Note: there is also *cscope* which is free, and maybe
        others. But as the name implies *cscope* is limited to C-like
        languages.

    *   Another corollary is that the tool must track each advance in each
        language, usually with significant lag time, and may not be
        bug-for-bug compatible with the compiler.

    *   Also, since analysis basically entails compiling the code, analysis
        of a large code base can take a long time, potentially as long or
        longer than actually building it.

    *   If some part of the application is written in a language the tool
        doesn't know (say Python or Visual Basic or Perl or an IDL), no
        analysis of that area can take place.

  PLUSES
    *   The analysis can be as granular and as language-knowledgeable as its
        developers can make it. If you change the signature of a C function,
        it can tell you how many uses of that function, in what files and on
        what lines, will need to change.

    *   A code analysis tool may be tied to a set of languages but by the
        same token it's NOT tied to a particular SCM or build system.

    The minuses above are not design flaws but inherent tradeoffs. For true
    code impact analysis you must buy one of these tools and accept the
    costs.

    Whouses doesn't attempt code analysis per se. As noted above, true code
    analysis programs are tied to language but not to an SCM system. Whouses
    flips this around; it doesn't care about language but it only works with
    build systems that use clearmake within a ClearCase VOB.

    Whouses takes the *config records* generated by clearmake, analyzes
    them, and tells you which files depend on which other files according to
    the CRs. Both techniques have fuzziness of different kinds: code
    analysis predicts what the real compiler will do based on what the
    analysis compiler found; divergence is possible. Whouses predicts what
    the next build will do based on what the last build did. If changes have
    taken place since, divergence is possible here too.

AUTHOR
    David Boyce <dsbperl AT boyski.com>

COPYRIGHT
    Copyright (c) 2000-2001 David Boyce. All rights reserved. This Perl
    program is free software; you may redistribute and/or modify it under
    the same terms as Perl itself.

STATUS
    This is currently ALPHA code and thus I reserve the right to change the
    UI incompatibly. At some point I'll bump the version suitably and remove
    this warning, which will constitute an (almost) ironclad promise to
    leave the interface alone.

PORTING
    I've tried to write this in a platform independent style but it hasn't
    been heavily tested on Windows (actually it hasn't been *all* that
    heavily tested anywhere). It does pass "make test" on Windows and
    appears to work fine in limited testing.

SEE ALSO
    perl(1), ClearCase::CRDB(3)



Changes
Revision history for Perl extension ClearCase::CRDB.

0.01 Tue Sep  5 11:48:14 EDT 2000
	- original version

0.02 Mon Oct 15 20:53:05 EDT 2001
	- Doc cleanup. No functional change.

0.03 Mon Oct 15 23:40:44 EDT 2001
	- Still tweaking docs.

0.04 Tue Oct 16 11:09:28 EDT 2001
	- More doc work.

0.05 Wed Oct 17 13:52:21 EDT 2001

0.06 Thu Oct 18 13:53:57 EDT 2001
	- Slight reworking of object layout; allow for multiple
	  iwd's in case of recursive builds.
	- Added recursive-make sample makefile and recmake.t test.

0.07 Sat Oct 27 21:37:12 EDT 2001
	- Added 'schema' number to deal with potential future reorgs.

0.08 Tue Jan  8 08:16:12 EST 2002
	- Allow Perl 5.005 - said to pass tests by a user.

0.09 Fri Nov 22 14:09:55 EST 2002
	- Fix test to insist on running within a view context.

0.10 Tue Jul 22 10:48:02 EDT 2003
	- Fixed contact data.

0.11 Thu Nov 10 07:59:11 EST 2005
	- Dealt with a slight change in CR format introduced in CC 6.0.
	  This simply suppresses a warning issued with the new format.

0.12 Mon Nov 21 16:53:48 EST 2005
	- Fixed a number of Windows-specific bugs involving paths
	  (/-vs-\, case sensitivity, and drive letters).
	- Fix: do not assume a variables section between MVFS Objects
	  and Build Script.


lib/ClearCase/CRDB/Denter.pm
package ClearCase::CRDB::Denter;

require ClearCase::CRDB;
require Data::Denter;

@ISA = qw(ClearCase::CRDB);

use strict;

sub store {
    my $self = shift;
    my $file = shift;
    my $d = Data::Denter->new;
    open(DUMP, ">$file") || die "$file: $!\n";
    printf DUMP "# Produced by Perl module %s using %s format.\n",
							    ref $self, ref $d;
    print DUMP "# This is a human-readable text format but not valid Perl\n";
    print DUMP $d->indent($self);
    close(DUMP);
    return $self;
}

sub load {
    my $self = shift;
    my $d = Data::Denter->new;
    for my $db (@_) {
	die "Error: $db: incorrect format" if -e $db && -B $db;
	open(DUMP, $db) || die "$db: $!\n";
	local $/ = undef;
	my $dented = <DUMP>;
	close(DUMP);
	my $hashref = $d->undent($dented);
	if (!defined($hashref)) {
	    warn "Error: $db: " . (-r $db ? $@ : $!);
	    return undef;
	}
	$self->hash_merge($hashref, @{$hashref->{CRDB_CRDO}});
    }
    return $self;
}

1;

__END__

=head1 NAME

ClearCase::CRDB::Denter - Data::Denter format subclass of ClearCase::CRDB

=head1 SYNOPSIS

Same as base class.

=head1 DESCRIPTION

This subclass of ClearCase::CRDB overrides the
I<ClearCase::CRDB-E<gt>load> and I<ClearCase::CRDB-E<gt>store> methods
to use the Data::Denter format rather than the Data::Dumper format.  I
know of no major advantage to Denter over Dumper in this application,
though that's not to say there aren't any ... I only added this
subclass to (a) test the extensibility of the base class and (b) learn
a bit more about Data::Denter.

=head1 AUTHOR

David Boyce <dsbperl AT boyski.com>

=head1 COPYRIGHT

Copyright (c) 2001 David Boyce. All rights reserved.  This Perl
program is free software; you may redistribute and/or modify it under
the same terms as Perl itself.

=head1 SEE ALSO

ClearCase::CRDB(3)


lib/ClearCase/CRDB/Dumper.pm
package ClearCase::CRDB::Dumper;

require ClearCase::CRDB;
require Data::Dumper;

@ISA = qw(ClearCase::CRDB);

use strict;

1;

__END__

=head1 NAME

ClearCase::CRDB::Dumper - text-format subclass of ClearCase::CRDB

=head1 SYNOPSIS

=head1 DESCRIPTION

This is an empty subclass of ClearCase::CRDB provided for consistency.
The native storage format of ClearCase::CRDB is that of Data::Dumper,
so ClearCase::CRDB::Dumper need not override the -E<gt>load and
-E<gt>store methods. However, other storage formats have their own
subclasses (cf ClearCase::CRDB::Storable) so ClearCase::CRDB::Dumper is
provided too. It has the same semantics as its base class.

=head1 AUTHOR

David Boyce <dsbperl AT boyski.com>

=head1 COPYRIGHT

Copyright (c) 2001 David Boyce. All rights reserved.  This Perl
program is free software; you may redistribute and/or modify it under
the same terms as Perl itself.

=head1 SEE ALSO

ClearCase::CRDB(3)


lib/ClearCase/CRDB/Storable.pm
package ClearCase::CRDB::Storable;

require ClearCase::CRDB;
require Storable;

@ISA = qw(ClearCase::CRDB);

use strict;

sub store {
    my $self = shift;
    my $file = shift;
    Storable::store($self, $file);
    return $self;
}

sub load {
    my $self = shift;
    for my $db (@_) {
	my $hashref;
	die "Error: $db: incorrect format" if -e $db && ! -B $db;
	# Binary: must have been created by Storable::store
	$hashref = Storable::retrieve($db);
	if (!defined($hashref)) {
	    warn "Error: $db: " . (-r $db ? $@ : $!);
	    return undef;
	}
	$self->hash_merge($hashref, @{$hashref->{CRDB_CRDO}});
    }
    return $self;
}

1;

__END__
=head1 NAME

ClearCase::CRDB::Storable - binary-format subclass of ClearCase::CRDB

=head1 SYNOPSIS

Same as base class.

=head1 DESCRIPTION

This subclass of ClearCase::CRDB overrides the
I<ClearCase::CRDB-E<gt>load> and I<ClearCase::CRDB-E<gt>store> methods
to use the Storable format rather than the Data::Dumper format.
Storable is believed to be faster both to write and read, but the
storage format is binary and thus not (easily) human readable.

=head1 AUTHOR

David Boyce <dsbperl AT boyski.com>

=head1 COPYRIGHT

Copyright (c) 2001 David Boyce. All rights reserved.  This Perl
program is free software; you may redistribute and/or modify it under
the same terms as Perl itself.

=head1 SEE ALSO

ClearCase::CRDB(3)


lib/ClearCase/CRDB.pm
package ClearCase::CRDB;

$VERSION = '0.12';

# This schema version is stored in the flat-file form via ->store
# and compared during ->load. A warning is issued if they don't match.
my $mod_schema = 1;

# I don't know exactly what Perl version is the minimum required but I
# have a report that it works with 5.005. I use 5.6.1 myself.
require 5.005;

use File::Spec 0.82;

# For convenience, load up known subclasses. We 'eval' these to ignore
# the error msg if the underlying serialization modules
# Data::Dumper/Storable/etc aren't installed.
eval { require ClearCase::CRDB::Dumper; };
eval { require ClearCase::CRDB::Denter; };
eval { require ClearCase::CRDB::Storable; };

use strict;

use constant MSWIN => $^O =~ /MSWin32|Windows_NT/i ? 1 : 0;

sub new {
    my $proto = shift;
    my($class, $self);
    if ($class = ref($proto)) {
	# Make a (deep) clone of the original
	require Clone;
	Clone->VERSION(0.11);	# there's a known bug in 0.10
	return Clone::clone($proto);
    }
    $class = $proto;
    $self = {};
    bless $self, $class;
    $self->catcr(@_) if @_;
    return $self;
}

sub crdo {
    my $self = shift;
    if (@_) {
	push(@{$self->{CRDB_CRDO}}, map {File::Spec->rel2abs($_)} @_);
	if (MSWIN) {
	    s%^[a-z]:%%i for (@{$self->{CRDB_CRDO}});
	}
    }
    return @{$self->{CRDB_CRDO}};
}

sub check {
    my $self = shift;
    $self->crdo(@_) if @_;
    my @objects = $self->crdo;
    die "Error: no derived objects specified to check" unless @objects;
    return system(qw(cleartool catcr -check -union), @objects);
}

sub catcr {
    my $self = shift;
    $self->crdo(@_) if @_;
    my($tgt, $state, @notes);
    my($Notes, $Objects, $Vars, $Script) = 1..4;
    open (CR, "cleartool catcr -r @{[$self->crdo]} |") || exit 2;
    for (<CR>) {
	chomp;
	next if /^-{28}/;

	# State machine - set $state according to the CR section.
        if (m%^Derived object:%) {
	    $state = $Notes;
	    next;
        } elsif (m%^Target\s+(\S*)\s+built\s%) {
	    $tgt = $1;
	    $tgt =~ s%^[a-z]:%%i if MSWIN;
	    $state = $Notes;
	    next;
	} elsif (($state == $Notes || $state == $Objects) && m%MVFS objects:%) {
	    @{$self->{CRDB_FILES}->{$tgt}->{CR_DO}->{CR_NOTES}} = @notes;
	    @notes = ();
	    $state = $Objects;
	    next;
	} elsif ($state == $Objects && m%^Variables and Options:%) {
	    $state = $Vars;
	    next;
	} elsif (($state == $Vars || $state == $Objects) && m%^Build Script:%) {
	    $state = $Script;
	    next;
	}

	# Accumulate data from section according to $state.
	if ($state == $Notes) {
	    push(@notes, $_);
	    if (my($base) = m%^Initial working directory was (\S+)%) {
		my $full = File::Spec->rel2abs($tgt, $base);
		$full =~ s%^[a-z]:%%i if MSWIN;
		$self->iwd($full, $base);
		if (-e $full) {
		    $tgt = $full;
		} else {
		    $self->{CRDB_FILES}->{$tgt}->{CR_PHONY} = 1;
		}
	    }
	} elsif ($state == $Objects) {
	    my($path, $vers, $date);
	    if (($path, $vers, $date) = m%^([/\\].+)@@(\S+)\s+<(\S+)>$%) {
		$path =~ s%^[a-z]:%%i if MSWIN;
		$self->{CRDB_FILES}->{$path}->{CR_TYPE} = 'ELEM';
		$self->{CRDB_FILES}->{$path}->{CR_ELEM}->{CR_VERS} = $vers;
		$self->{CRDB_FILES}->{$path}->{CR_DATE} = $date;
	    } elsif (($path, $date) = m%^([/\\].+)@@(\S+)$%) {
		$path =~ s%^[a-z]:%%i if MSWIN;
		$self->{CRDB_FILES}->{$path}->{CR_TYPE} = 'DO';
		$self->{CRDB_FILES}->{$path}->{CR_DATE} = $date;
	    } elsif (($path, $date) = m%^([/\\].+\S)\s+<(\S+)>$%) {
		$path =~ s%^[a-z]:%%i if MSWIN;
		$self->{CRDB_FILES}->{$path}->{CR_TYPE} = 'NON';
		$self->{CRDB_FILES}->{$path}->{CR_DATE} = $date;
	    } else {
		warn "Warning: unrecognized CR line: '$_'";
		next;
	    }
	    if (MSWIN) {
		# Must compare paths case-insensitively on Windows.
		next if lc($path) eq lc($tgt);
	    } else {
		next if $path eq $tgt;
	    }
	    $self->{CRDB_FILES}->{$tgt}->{CR_DO}->{CR_NEEDS}->{$path} = 1;
	    next if exists $self->{CRDB_FILES}->{$tgt}->{CR_PHONY};
	    $self->{CRDB_FILES}->{$path}->{CR_MAKES}->{$tgt} = 1;
	} elsif ($state == $Vars && (my($var, $val) = m%^(.+?)=(.*)%)) {
	    $self->{CRDB_FILES}->{$tgt}->{CR_DO}->{CR_VARS}->{$var} = $val;
	} elsif ($state == $Script) {
	    push(@{$self->{CRDB_FILES}->{$tgt}->{CR_DO}->{CR_SCRIPT}},
								substr($_, 1));
	} else {
	    warn "Warning: unrecognized CR line: '$_'";
        }
    }
    close (CR);
    return $self;
}

# Internal func to recursively merge two hash refs
sub hash_merge {
    my($to, $from, @source) = @_;
    for (keys %{$from}) {
	if (! $to->{$_}) {
	    $to->{$_} = $from->{$_};
	    next;
	}
	my($ttype, $ftype) = (ref $to->{$_}, ref $from->{$_});
	if ($ttype ne $ftype) {
	    warn "Warning: key type conflict: @source: $_ $ttype/$ftype"
	} elsif (! $ttype) {
	    warn "Warning: @source: $_: can't merge non-references";
	} elsif ($ttype eq 'ARRAY') {
	    push(@{$to->{$_}}, @{$from->{$_}});
	} elsif ($ttype eq 'HASH') {
	    hash_merge($to->{$_}, $from->{$_}, @source);
	} else {
	    warn "Warning: @source: $_: can't merge type: $ttype";
	}
    }
}

sub store {
    my $self = shift;
    my $file = shift;
    $self->{CRDB_SCHEMA} = $mod_schema;
    my $d = Data::Dumper->new([$self], ['_DO']);
    $d->Indent(1);
    open(DUMP, ">$file") || die "$file: $!\n";
    printf DUMP "# Produced by Perl module %s using %s format.\n",
							    ref $self, ref $d;
    print DUMP "# It is valid Perl syntax and may be read into memory via\n";
    print DUMP "# 'do <file>' or by eval-ing its contents.\n\n";
    print DUMP $d->Dumpxs;
    close(DUMP);
    return $self;
}

sub load {
    my $self = shift;
    for my $db (@_) {
	my $hashref;
	die "Error: $db: incorrect format" if -e $db && -B $db;
	$hashref = do $db;	# eval's $db and returns the obj ref
	if (!defined($hashref)) {
	    warn "Error: $db: " . (-r $db ? $@ : $!);
	    return undef;
	}
	my $file_schema = $hashref->{CRDB_SCHEMA};
	die "Error: $db: stored schema ($file_schema) != current ($mod_schema)"
					unless $file_schema == $mod_schema;
	$self->hash_merge($hashref, @{$hashref->{CRDB_CRDO}});
    }
    return $self;
}

sub iwd {
    my $self = shift;
    if (@_ == 2) {
	my($path, $iwd) = @_;
	$self->{CRDB_FILES}->{$path}->{CR_DO}->{CR_IWD} = $iwd;
    } elsif (@_ == 1) {
	my($path) = @_;
	return $self->{CRDB_FILES}->{$path}->{CR_DO}->{CR_IWD};
    } elsif (@_ == 0) {
	my %iwds;
	for (keys %{$self->{CRDB_FILES}}) {
	    $iwds{$_} = $self->{CRDB_FILES}->{$_}->{CR_DO}->{CR_IWD};
	}
	return sort keys %iwds;
    }
    return undef;
}

sub files {
    my $self = shift;
    return keys %{$self->{CRDB_FILES}};
}

sub targets {
    my $self = shift;
    return grep {exists $self->{CRDB_FILES}->{$_}->{CR_DO}}
						    keys %{$self->{CRDB_FILES}};
}

sub vars {
    my $self = shift;
    my $do = shift;
    return keys %{$self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_VARS}};
}

sub val {
    my $self = shift;
    my($do, $var) = @_;
    return undef unless exists $self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_VARS};
    return $self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_VARS}->{$var};
}

sub phony {
    my $self = shift;
    if (@_) {
	return exists $self->{CRDB_FILES}->{$_[0]}->{CR_PHONY};
    } else {
	return grep {exists $self->{CRDB_FILES}->{$_}->{CR_PHONY}}
						    keys %{$self->{CRDB_FILES}};
    }
}

sub macroize {
    my $self = shift;
    my @dos = @_ ? @_ : keys %{$self->{CRDB_FILES}};
    my %used;
    for my $do (@dos) {
	next unless exists $self->{CRDB_FILES}->{$do}->{CR_DO} &&
		    exists $self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_SCRIPT} &&
		    exists $self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_VARS};
	my %macros = %{$self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_VARS}};
	for (@{$self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_SCRIPT}}) {
	    for my $var (keys %macros) {
		next unless $macros{$var};
		if (length($var) == 1) {
		    $used{$var} = $macros{$var} if s%\Q$macros{$var}%\$$var%g;
		} else {
		    $used{$var} = $macros{$var} if s%\Q$macros{$var}%\$($var)%g;
		}
	    }
	}
    }
    return %used;
}

sub notes {
    my $self = shift;
    my $do = shift;
    return undef unless exists $self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_NOTES};
    return @{$self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_NOTES}};
}

sub script {
    my $self = shift;
    my $do = shift;
    return undef
	    if !exists $self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_SCRIPT};
    return @{$self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_SCRIPT}};
}

sub matches_do {
    my $self = shift;
    my @matched;
    for my $re (@_) {
	push(@matched, grep m%$re%, keys %{$self->{CRDB_FILES}});
    }
    return @matched;
}

sub needs_do {
    my $self = shift;
    my @results;
    for my $do (@_) {
	push(@results, keys %{$self->{CRDB_FILES}->{$do}->{CR_DO}->{CR_NEEDS}})
				    if exists $self->{CRDB_FILES}->{$do};
    }
    return @results;
}

sub makes_do {
    my $self = shift;
    my @results;
    for my $do (@_) {
	push(@results, keys %{$self->{CRDB_FILES}->{$do}->{CR_MAKES}})
					if exists $self->{CRDB_FILES}->{$do};
    }
    return @results;
}

1;

__END__

=head1 NAME

ClearCase::CRDB - base class for ClearCase config-record analysis

=head1 SYNOPSIS

    my $crdb = ClearCase::CRDB->new(@ARGV);	# @ARGV is a list of DO's
    $crdb->check;				# Do a CR sanity check
    $crdb->catcr;				# Analyze the recursive CR
    $crdb->store($filename);			# Dump CR to $filename

=head1 DESCRIPTION

A ClearCase::CRDB object represents the recursive I<configuration
record> (aka I<CR>) of a set of I<derived objects> (aka I<DO's>).  It
provides methods for easy extraction of parts of the CR such as the
build script, MVFS files used in the creation of a given DO, make
macros employed, etc. This is the same data available from ClearCase in
raw textual form from "cleartool catcr -recurse DO ..."; it's just
broken down for easier access and analysis.

An example of what can be done with ClearCase::CRDB is the provided
I<whouses> script which, given a particular DO, can show recursively
which files it depends on or which files depend on it.

Since recursively deriving a CR database can be a slow process for
large build systems and can burden the VOB database, the methods
C<ClearCase::CRDB-E<gt>store> and C<ClearCase::CRDB-E<gt>load> are
provided. These allow the derived CR data to be stored in its processed
form to a persistent storage such as a flat file or database and
re-loaded from there. For example, this data might be derived once per
day as part of a nightly build process and would then be available for
use during the day without causing additional VOB load.

The native C<ClearCase::CRDB-E<gt>store> and
C<ClearCase::CRDB-E<gt>load> methods store to a flat file in
human-readable text format. Different formats may be used by
subclassing these two methods. An example subclass
C<ClearCase::CRDB::Storable> is provided; this uses the Perl module
I<Storable> which is a binary format. If you wanted to store to a
database this is how you'd do it.

=head2 CONSTRUCTOR

Use C<ClearCase::CRDB-E<gt>new> to construct a CRDB object. Any
parameters given will be taken as the set of derived objects to
analyze.

=head2 INSTANCE METHODS

Following is a brief description of each supported method. Examples
are given for all methods that take parameters; if no example is
given usage may be assumed to look like:

    my $result = $obj->method;

Also, if the return value is described in plural terms it may be
assumed that the method returns a list.

=over 4

=item * crdo

Sets or gets the list of derived objects under consideration, e.g.:

    $obj->crdo(qw(do_1, do_2);	# give the object a list of DO's
    my @dos = $obj->crdo;	# gets the list of DO's

This method is invoked automatically by the constructor (see) if
derived objects are specified.

=item * catcr

Invokes I<cleartool catcr -recurse> on the DO set and breaks the
resultant textual data apart into various fields which may then be
accessed by the methods below. This method is invoked automatically by
the constructor (see) if derived objects are specified.

=item * check

Checks the set of derived objects for consistency. For instance, it
checks for multiple versions of the same element, or multiple
references to the same element under different names, in the set of
config records.

=item * store

Writes the processed config record data to the specified file.

=item * load

Reads processed config record data from the specified files.

=item * needs_do

Takes a list of derived objects, returns the list of derived objects
which they use. For example, if C<foo.c> includes C<foo.h> and compiles
to C<foo.o> which then links to the executable C<foo>, the
C<-E<gt>needs_do> method when given C<foo.o> would return the list
C<('foo.c', 'foo.h')>. In other words it returns "upstream
dependencies".

=item * makes_do

Takes a list of derived objects, returns the list of derived objects
which use them. This is the reverse of C<needs_do>. Given the
C<needs_do> example above, the C<-E<gt>makes_do> method when given
C<foo.o> would return C<foo>. In other words it returns "downstream
dependencies".

=item * iwd

Each target in a CR has an "initial working directory" or I<iwd>. If
passed a DO, this method returns the I<iwd> of that derived object.
With no parameters it returns the list of I<iwds> mentioned in the CR.

=item * files

Returns the complete set of files mentioned in the CR.

=item * targets

Returns the subset of files mentioned in the CR which are targets.

=item * vars

Returns the set of make macros used in the build script for the
specified DO, e.g.:

    my @list = $obj->vars("path-to-derived-object");

=item * val

Returns the value of the specified make macro as used in the build script
for the specified DO:

    my $value = $obj->val("path-to-derived-object", "CC");

=item * notes

Returns the set of "build notes" for the specified DO as a list. This
is the section of the CR which looks like:

    Target foo built by ...
    Host "host" running ...
    Reference Time ...
    View was ...
    Initial working directory was ...

E.g.

    my @notes = $obj->notes("path-to-derived-object");

=item * script

Returns the build script for the specified DO:

    my $script = $obj->script("path-to-derived-object");

=back

There are also some undocumented methods in the source. This is
deliberate; they're experimental.

=head1 AUTHOR

David Boyce <dsbperl AT cleartool.com>

=head1 COPYRIGHT

Copyright (c) 2000-2005 David Boyce. All rights reserved.  This Perl
program is free software; you may redistribute and/or modify it under
the same terms as Perl itself.

=head1 STATUS

This is currently ALPHA code and thus I reserve the right to change the
API incompatibly. At some point I'll bump the version suitably and
remove this warning, which will constitute an (almost) ironclad promise
to leave the interface alone.

=head1 PORTING

This module has been at least slightly tested, at various points in its
lifecycle, on almost all CC platforms including Solaris 2.6-8, HP-UX 10
and 11, and Windows NT4 and Win2K SP2 using perl 5.004_04 through 5.6.1
and CC4.1 through 5.0.  However, I tend to use the latest of everything
(CC5.0, Solaris8, Win2KSP2, Perl5.6.1 at this writing) and cannot
regression-test with anything earlier. Also, note that I rarely use
this on Windows so it may be buggier there.

=head1 BUGS

NOTE: A bug in CC 5.0 causes CRDB's "make test" to dump core. This bug
is in clearmake, not CRDB, and in any case affects only its test
suite.  The first CC 5.0 patch contains a fix, so you probably don't
want to use CC 5.0 unpatched. If you do, ignore the core dump in
the test suite and force the install anyway.

Please send bug reports or patches to the address above.

=head1 SEE ALSO

perl(1), ct+config_record(1), clearmake(1) et al
